### PR TITLE
API: Implemented sync plan tests

### LINKED ISSUE #2445: API Automate update sync plan
please automate: update sync plan

PUT /katello/api/organizations/:organization_id/sync_plans/:id
Update a sync plan
PUT /katello/api/sync_plans/:id
Update a sync plan
Params
Param name 	Description
organization_id
optional 	

Filter sync plans by organization name or label

Value: number.
id
required 	

sync plan numeric identifier

Value: number.
name
optional 	

sync plan name

Value: String
interval
optional 	

how often synchronization should run

Value: Must be one of: hourly, daily, weekly.
sync_date
optional 	

start datetime of synchronization

Value: String
description
optional 	

sync plan description

Value: String
enabled
optional 	

enables or disables synchronization

### LINKED ISSUE #2446: API  Automate delete sync plan
please automate delete sync plan

DELETE /katello/api/organizations/:organization_id/sync_plans/:id
Destroy a sync plan
DELETE /katello/api/sync_plans/:id
Destroy a sync plan
Params
Param name 	Description
organization_id
optional 	

Filter sync plans by organization name or label

Value: number.
id
optional 	

sync plan numeric identifier

Value: number.

### PR DESCRIPTION
**Pull request by abalakh**
Implemented stubbed sync plan tests and added few more for better coverage.
Closes #2445
Closes #2446
Tests results:
```
nosetests tests/foreman/api/test_syncplan.py
.............................................S.SSS.........................................
----------------------------------------------------------------------
Ran 91 tests in 177.606s

OK (SKIP=4)
```
`[DO NOT MERGE]` as it depends on SatelliteQE/nailgun#152

### CODE REVIEW

File: robottelo/common/helpers.py [hunk 1]
```diff
@@ -229,7 +229,7 @@ def invalid_values_list():
     code.
 
     """
-    return ['', ' ', '   '] + invalid_names_list()
+    return ['', ' ', '\t'] + invalid_names_list()
```
Comment (abalakh) [inline robottelo/common/helpers.py]: Supposed to be a tab symbol here. Probably IDE autoreplaced it with 3 spaces. Anyway, `\t` should work for sure.

File: tests/foreman/api/test_syncplan.py [hunk 2]
```diff
@@ -167,9 +175,52 @@ def test_create_sync_plan_with_invalid_random_name(self, name):
...
 
-@stubbed()
+        @Assert: A sync plan can not be created with empty 'enabled' field.
+
+        @Feature: SyncPlan
+
+        """
+        with self.assertRaises(HTTPError):
+            entities.SyncPlan(
+                enabled=enabled,
```
Comment (Ichimonji10) [inline tests/foreman/api/test_syncplan.py]: Is the intent in this test to provide an invalid value for 'enabled', or to provide no value at all? It's possible to simply omit the "enabled" field like so:

```python
entities.SyncPlan(
    interval=…,
    name=…,
    organization=…,
    sync_date=…,
).create(False)  # create_missing=False
```

Comment (abalakh) [inline tests/foreman/api/test_syncplan.py]: I'm checking both no value and invalid value here. In both cases sync_plan creation should fail.
What's the advantage of omitting a field in that way? We will have to manually specify some random interval, name, generate some date. It won't decrease code quantity, and, won't increase code readability - imo, if only 'enabled' parameter specified means i'm working only with it; if 5 arguments are specified but one is missing - i'm not sure which one is missing and what i'm actually testing here.
Everything above is just imo, please, share your thoughts on that :)

File: tests/foreman/api/test_syncplan.py [hunk 3]
```diff
@@ -258,9 +295,19 @@ def test_update_sync_plan_with_random_interval(self, interval):
         @Feature: SyncPlan
 
         """
+        initial_interval = [
```
Comment (Ichimonji10) [inline tests/foreman/api/test_syncplan.py]: I think this has the same effect.

```python
sync_plan = entities.SyncPlan(organization=self.org)
sync_plan.interval = choice(  # from random import choice
    set(sync_plan.get_fields()['interval'].choices) - set([interval])
)
sync_plan = sync_plan.create()
sync_plan.interval = interval
self.assertEqual(sync_plan.update(['interval']).interval, interval)
```

File: tests/foreman/api/test_syncplan.py [hunk 4]
```diff
@@ -274,14 +321,78 @@ def test_update_sync_plan_with_random_interval(self, interval):
...
+    @data(*invalid_values_list())
+    def test_update_sync_plan_with_invalid_name_negative(self, name):
+        """@Test: Update a sync plan with an invalid name.
+
+        @Assert: A sync plan can not be updated with the specified name.
+
+        @Feature: SyncPlan
+
+        """
+        syncplan = entities.SyncPlan(
```
Comment (Ichimonji10) [inline tests/foreman/api/test_syncplan.py]: Nitpick: all on one line.

File: tests/foreman/api/test_syncplan.py [hunk 5]
```diff
@@ -274,14 +321,78 @@ def test_update_sync_plan_with_random_interval(self, interval):
...
+    )
+    def test_update_sync_plan_with_empty_enabled_value_negative(self, enabled):
+        """@Test: Update a sync plan with empty 'enabled' field.
+
+        @Assert: A sync plan can not be updated with empty 'enabled' field.
+
+        @Feature: SyncPlan
+
+        """
+        syncplan = entities.SyncPlan(
```
Comment (Ichimonji10) [inline tests/foreman/api/test_syncplan.py]: Nitpick: all on one line. (What can I say - a four line unit test makes me happy. :)

File: tests/foreman/api/test_syncplan.py [hunk 6]
```diff
@@ -221,17 +257,13 @@ def test_update_sync_plan_with_random_name(self, name):
         @Feature: SyncPlan
 
         """
+        syncplan = entities.SyncPlan(
```
Comment (Ichimonji10) [inline tests/foreman/api/test_syncplan.py]: Nitpick: all on one line.

File: tests/foreman/api/test_syncplan.py [hunk 7]
```diff
@@ -274,14 +321,78 @@ def test_update_sync_plan_with_random_interval(self, interval):
...
+        """@Test: Update a sync plan with empty interval specified.
 
-        @Assert: A sync plan can be created and its sync date can be updated
-        with the specified sync date.
+        @Assert: A sync plan can not be updated with empty interval specified.
 
         @Feature: SyncPlan
 
         """
+        syncplan = entities.SyncPlan(
```
Comment (Ichimonji10) [inline tests/foreman/api/test_syncplan.py]: Nitpick: all on one line.

File: tests/foreman/api/test_syncplan.py (hunk 2 above)
Comment (Ichimonji10) [inline tests/foreman/api/test_syncplan.py]: > What's the advantage of omitting a field in that way?

It depends on what you want to test. Do you want to test *invalid* values or *missing* values?

> We will have to manually specify some random interval, name, generate some date. […]

See second example, above.

File: tests/foreman/api/test_syncplan.py [hunk 8]
```diff
@@ -481,3 +600,11 @@ def test_delete_sync_plan_with_one_synced_product(self):
         @Feature: SyncPlan
 
         """
+        syncplan = entities.SyncPlan(organization=self.org).create()
+        product = entities.Product(organization=self.org).create()
+        entities.Repository(product=product).create()
+        syncplan.add_products([product.id])
+        product.sync()
```
Comment (sthirugn) [inline tests/foreman/api/test_syncplan.py]: Is this a custom product? Where is this getting the sync data from?

File: tests/foreman/api/test_syncplan.py (hunk 6 above)
Comment (abalakh) [inline tests/foreman/api/test_syncplan.py]: Can't - resulting line will be 86 chars long. But thanks for pointing, will double check everything and merge into one line where it's possible.

Comment (Ichimonji10) [inline tests/foreman/api/test_syncplan.py]: Whoops!

File: robottelo/common/helpers.py (hunk 1 above)
Comment (Ichimonji10) [inline robottelo/common/helpers.py]: :+1:

File: tests/foreman/api/test_syncplan.py (hunk 3 above)
Comment (abalakh) [inline tests/foreman/api/test_syncplan.py]: And it is much better, as we're avoiding predefined values. Thanks for suggestion :+1:

File: tests/foreman/api/test_syncplan.py (hunk 8 above)
Comment (abalakh) [inline tests/foreman/api/test_syncplan.py]: I might be wrong, as I'm not very familiar with this functionality, but as I understand synchronize command for product actually synchronizes all repositories in it. 
Repositories, created with nailgun by default are of 'yum' type with [following URL](https://github.com/SatelliteQE/nailgun/blob/master/nailgun/entities.py#L64):
```python
_FAKE_YUM_REPO = 'http://inecas.fedorapeople.org/fakerepos/zoo3/'
```
That's probably the answer to your question. Please, correct me if i'm wrong in smth.

**Comment by abalakh** [comment]
As per IRC discussion:
1) 'enabled' field appeared to be optional, so there's no point in tests validating no input for it.
2) E.g. `entities.SyncPlan(organization=None).create()` will send `organization` parameter (equal to 'None'). To avoid sending some parameter - it shouldn't be specified at all and `create_missing` should be disabled.

Updated tests according to this points and comments. Also made a rebase.
```
nosetests tests/foreman/api/test_syncplan.py                                             
..................................................S.SSS.............................................
----------------------------------------------------------------------
Ran 100 tests in 184.208s

OK (SKIP=4)
```

File: tests/foreman/api/test_syncplan.py (hunk 8 above)
Comment (sthirugn) [inline tests/foreman/api/test_syncplan.py]: yeah ok, thank you.

**Comment by sthirugn** [comment]
Removed `Do Not Merge` prefix since the nailgun pr is merged

File: tests/foreman/api/test_syncplan.py [hunk 9]
```diff
@@ -167,9 +172,40 @@ def test_create_sync_plan_with_invalid_random_name(self, name):
...
+        @Assert: A sync plan can not be created with no interval specified.
+
+        @Feature: SyncPlan
+
+        """
+        with self.assertRaises(HTTPError):
+            sync_plan = entities.SyncPlan(organization=self.org)
+            sync_plan.create_missing()
+            del sync_plan.interval
+            sync_plan.create(False)
```
Comment (Ichimonji10) [inline tests/foreman/api/test_syncplan.py]: I think only this line of code needs to be in the `with self.assertRaises` context manager.

File: tests/foreman/api/test_syncplan.py [hunk 10]
```diff
@@ -258,9 +278,15 @@ def test_update_sync_plan_with_random_interval(self, interval):
         @Feature: SyncPlan
 
         """
+        sync_plan = entities.SyncPlan(organization=self.org)
+        sync_plan.interval = sample(
+            set(sync_plan.get_fields()['interval'].choices) - set([interval]),
+            1
+        )[0]
```
Comment (Ichimonji10) [inline tests/foreman/api/test_syncplan.py]: FYI, this has the same result:

```python
sync_plan.interval = choice(tuple(
    set(sync_plan.get_fields()['interval'].choices) - set([interval])
))
```

What you have is fine, though.

**Comment by Ichimonji10** [comment]
ACK pending comment re: context manager placement.

**Comment by abalakh** [comment]
Updated context manager placement & made a rebase :)

**Comment by Ichimonji10** [comment]
:+1:

**Comment by sthirugn** [comment]
ACK

**Pull request merged**